Add osg-test CI

brianhlin opened the pull request:
This PR adds CI that:

1. Builds the blahp based on the `condor` RPM available in the 8.9 CHTC and 3.5 upcoming development OSG Yum repositories (both 8.9.13 at the time of writing)
1. Runs OSG integration tests against PBS and Slurm batch systems

Currently the tests are failing due to https://opensciencegrid.atlassian.net/browse/HTCONDOR-395

brianhlin committed:
First pass at osg-test CI

brianhlin committed:
Grab Slurm logs on test failure

brianhlin committed:
Set hostname on container startup for HTCondor and Slurm

brianhlin committed:
Generate user proxies for job submission

brianhlin committed:
Update custom condor configs from htcondor/htcondor-ce

brianhlin committed:
Jobs get routed to local HTCondor in the osg-test setup

So we need the logs in case of failures

brianhlin committed:
Speed up Job Router/Gridmanager queries

brianhlin committed:
Disable BLAHP status debugging

Status commands will bail almost immediately if the cache dir is not
owned by the appropriate user. However, creating and chown'ing the dir
beforehand won't work because then osg-test will skip creating the
user cert, causing early test failures

brianhlin committed:
Revert "Disable BLAHP status debugging"

This reverts commit 738bfe2a05d856333787eb724c813730edf51d1a. We can
do this because we separated out the user cert generation from the
user creation in osg-test

brianhlin committed:
Re-enable blahp status cache debug logs

brianhlin committed:
Give PBS/Torque jobs more time to run

brianhlin committed:
Build the BLAHP against HTCondor available in the CHTC Yum repo

brianhlin committed:
Install the OSG Yum repos so we can install the osg-ce metapackages

brianhlin committed:
Add test job for the OSG environment

brianhlin committed:
Use find for dumping logs; cat fails if there are subdirs

edquist commented on .github/workflows/build_and_test_rpms.yml:
```diff
@@ -0,0 +1,103 @@
+                 --hostname "blahp-ci.htcondor.org" \
+                 --name $GITHUB_SHA \
+                 --volume /sys/fs/cgroup:/sys/fs/cgroup \
+                 --volume `pwd`:/blahp:rw \
```
```suggestion
                 --volume "$PWD":/blahp:rw \
```

edquist commented on .github/workflows/build_and_test_rpms.yml:
```diff
@@ -0,0 +1,103 @@
+                 /bin/bash -xc \
+                   "/blahp/tests/install_rpms.sh \
+                     ${{ matrix.dver }} \
+                     ${{ matrix.target_env}}"
```
What is going on here with `-xc` ?  Do you intend for the `-x` to apply to the commands in `install_rpms.sh`, or just to the `install_rpms.sh` command line, in order to see what dver / target_env get used?  (Because i think in the case of `-xc`, the `-x` only applies to the `-c` command line itself.)

edquist commented on .github/workflows/build_and_test_rpms.yml:
```diff
@@ -0,0 +1,103 @@
+                     ${{ matrix.target_env}}"
+
+      - name: Host certificate details
+        if: failure() && steps.integration-tests.outcome == 'failure'
```
How does the `failure()` function work here? Is it defined somewhere or is it a standard workflow thing?

(I notice the same condition is used for a number of these sections, and i am somewhat curious if `failure()` is evaluated each time / if the intermediate runs affect its results.)

brianhlin replied:
Yup, it's a built-in that returns true whenever any previous job step fails: https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#failure

edquist commented on tests/setup_tests.sh:
```diff
@@ -0,0 +1,27 @@
+ # Run tests in Container
+if [ "$el_version" = "6" ]; then
+
+sudo docker run --rm=true -v `pwd`:/blahp:rw centos:centos${OS_VERSION} /bin/bash -c "bash -xe /blahp/tests/test_inside_docker.sh ${OS_VERSION}"
```
Are `$el_version` and `$OS_VERSION` not the same thing?  Somewhat curious if they are redundant

edquist replied:
Minor, but...

```suggestion
if [ "$el_version" = "6" ]; then

sudo docker run --rm=true -v "$PWD":/blahp:rw centos:centos${OS_VERSION} /bin/bash -c "bash -xe /blahp/tests/test_inside_docker.sh ${OS_VERSION}"
```

edquist commented on tests/setup_tests.sh:
```diff
@@ -0,0 +1,27 @@
+
+elif [ "$el_version" = "7" ]; then
+
+docker run --privileged -d -ti -e "container=docker"  -v /sys/fs/cgroup:/sys/fs/cgroup -v `pwd`:/blahp:rw  centos:centos${OS_VERSION}   /usr/sbin/init
```
```suggestion
docker run --privileged -d -ti -e "container=docker"  -v /sys/fs/cgroup:/sys/fs/cgroup -v "$PWD":/blahp:rw  centos:centos${OS_VERSION}   /usr/sbin/init
```

edquist commented on tests/setup_tests.sh:
```diff
@@ -0,0 +1,27 @@
+elif [ "$el_version" = "7" ]; then
+
+docker run --privileged -d -ti -e "container=docker"  -v /sys/fs/cgroup:/sys/fs/cgroup -v `pwd`:/blahp:rw  centos:centos${OS_VERSION}   /usr/sbin/init
+DOCKER_CONTAINER_ID=$(docker ps | grep centos | awk '{print $1}')
```
Are you confident that this will always/only return exactly 1 running container matching the string `centos`?

edquist commented on tests/test_inside_docker.sh:
```diff
@@ -0,0 +1,50 @@
+chown -R vdttest: /var/tmp/*_cache_vdttest/
+
+# Ok, do actual testing
+set +e # don't exit immediately if osg-test fails
```
I don't get it.

>don't exit immediately if osg-test fails

_Instead_, save the return code from osg-test, and _then_ (immediately) exit with that code.

Seems like the same thing?

brianhlin replied:
Ah yeah, it's an artifact from when we would run `osg-test` then grab all of the log files in the same script. Now that we do the latter in separate job steps, we can drop this.

edquist commented on tests/test_inside_docker.sh:
```diff
@@ -0,0 +1,50 @@
+popd
+
+# Bind on the right interface and skip hostname checks.
+cat << EOF > /etc/condor/config.d/99-local.conf
+BIND_ALL_INTERFACES = true
+GSI_SKIP_HOST_CHECK=true
+ALL_DEBUG=\$(ALL_DEBUG) D_FULLDEBUG D_CAT
+SCHEDD_INTERVAL=1
+SCHEDD_MIN_INTERVAL=1
+JOB_ROUTER_POLLING_PERIOD=1
+GRIDMANAGER_JOB_PROBE_INTERVAL=1
+EOF
```
Fine how you are doing it, but note if you want to do a here-doc without variable expansion, you can just quote the delimiter word (rather than escaping the `$` signs), eg:

```suggestion
cat << "EOF" > /etc/condor/config.d/99-local.conf
BIND_ALL_INTERFACES = true
GSI_SKIP_HOST_CHECK=true
ALL_DEBUG=$(ALL_DEBUG) D_FULLDEBUG D_CAT
SCHEDD_INTERVAL=1
SCHEDD_MIN_INTERVAL=1
JOB_ROUTER_POLLING_PERIOD=1
GRIDMANAGER_JOB_PROBE_INTERVAL=1
EOF
```

brianhlin replied:
Neato!

edquist commented on .github/workflows/build_and_test_rpms.yml:
```diff
@@ -0,0 +1,103 @@
+        if: failure() && steps.integration-tests.outcome == 'failure'
+        run: |
+          docker exec $GITHUB_SHA \
+                 /bin/bash -c "find /var/log/condor-ce -type f -exec cat {} \;"
```
In general for these commands, is the idea just to test if the command runs successfully, or do you actually want to capture the output for later inspection?

If it's the latter, note that running `cat` on a bunch of files will not show which lines belong to which files in the output.  (You could either add a header/footer with additional items in the `find` command (eg with `-printf`), or else do something like `grep ^` instead of `cat`.)

Also, the `bash -c` thing here is fine, and is kind of necessary when you want a glob (`*`) to be evaluated inside the container rather than where the docker command line is invoked; it seems in the case of this `find` command line (with no wildcards), i _think_ you can also get away with just

```suggestion
                 find /var/log/condor-ce -type f -exec cat {} \;
```

(wild, right?)

Same goes for every other `docker exec ... bash -c "command line"`, where `command line` doesn't have any wildcards.

brianhlin replied:
We want the actual output for inspection in case of test failure. I've kicked off another branch that tries out your suggested change

brianhlin replied:
The change works. It would be nice to have separators between files but we're in a bit of a rush to get this in so I'm going to go ahead and merge

edquist commented on tests/build_rpms.sh:
```diff
@@ -0,0 +1,102 @@
+useradd -r -g condor -d /var/lib/condor -s /sbin/nologin \
+        -u 64 -c "Owner of HTCondor Daemons" condor
+
+if  [[ $OS_VERSION == 7 ]]; then
+    YUM_PKG_NAME="yum-plugin-priorities"
+else
+    YUM_PKG_NAME="yum-utils"
+fi
```
I don't suppose there is a virtual capability or path (like `/usr/bin/yum-config-manager`) that you could use instead of the package name here that would cover the thing you actually care about installing?

brianhlin replied:
¯\_(ツ)_/¯ maybe! That's something to put in our back pocket since that's what we're doing in [software-base](https://github.com/opensciencegrid/docker-software-base/blob/master/Dockerfile#L14-L22), too

edquist commented on tests/build_rpms.sh:
```diff
@@ -0,0 +1,102 @@
+    gzip > /tmp/rpmbuild/SOURCES/blahp-${package_version}.tar.gz
+popd
+
+# Build the SRPM; don't put a dist tag in it
+rpmbuild --define '_topdir /tmp/rpmbuild' --undefine 'dist' -bs /tmp/rpmbuild/SPECS/blahp.spec
+# Build the binary RPM
+rpmbuild --define '_topdir /tmp/rpmbuild' -bb /tmp/rpmbuild/SPECS/blahp.spec
```
What is your intention here in omitting the dist tag from the srpm? ... Somehow it feels a bit off to me. I guess i am just used to seeing the dist tag match between source and binary rpms.

brianhlin replied:
I grabbed this from the HTCondo-CE build/test CI, which we use to actually build the HTCondor-CE RPMs distributed in the CHTC Yum repos. IIRC, flightworthy didn't want the dist tags for the SRPMs? I could be making that up, though

brianhlin replied:
In light of the above and since the tests work, I'm going to go ahead and leave this unaddressed

edquist commented on tests/build_rpms.sh:
```diff
@@ -0,0 +1,102 @@
+    [[ $? -eq 0 ]] && break
+    sleep $((5**$retry))
+done
+set -e
```
Fine how you have it - but note that `set -e` does not trigger an exit on failure when the failed command is on the left side of a `&&` or `||`; so you can equivalently do something like

```suggestion
for retry in {1..3}; do
    yum -y -d0 update && break  # Update the OS packages
    sleep $((5**$retry))
done
```

edquist replied:
I would also note that if you unroll this loop and the yum update continues to fail, the command sequence is

```
yum -y -d0 update
sleep 5
yum -y -d0 update
sleep 25
yum -y -d0 update
sleep 125
```

The final sleep is not followed by a final retry (probably not really what you intended), and if the final retry fails, the script still continues (not sure whether that's what you had in mind or not).

There are different ways you could re-write it.  Here's one off the cuff:

```suggestion
retry=0
until yum -y -d0 update || ((++retry > 3)); do
    sleep $((5**$retry))
done
# maybe die if ((retry > 3))
```

edquist commented on tests/build_rpms.sh:
```diff
@@ -0,0 +1,102 @@
+    python3-rpm
+
+# Prepare the RPM environment
+mkdir -p /tmp/rpmbuild/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
```
not that it hurts to create these, but i thought rpmbuild does that for you

brianhlin replied:
Kicked off a branch with this removed. Let's find out if `rpmbuild` does the right thing!

brianhlin replied:
Had to still pre-create some dirs since we copy data into them but otherwise it looks like `rpmbuild` does what we want.

edquist reviewed (commented):
Eh, enough comments for now.

brianhlin committed:
Use PWD env var instead of calling out to pwd

Co-authored-by: edquist <edquist@users.noreply.github.com>

brianhlin committed:
Remove stale test script

brianhlin committed:
Remove vestigal test exit code handling

We used to grab debug info like logs after the test within this script
but now we do that in separate job steps

brianhlin committed:
Improve writing of HTCondor{,-CE} config

We can do this since we're not subbing any variables

Co-authored-by: edquist <edquist@users.noreply.github.com>

brianhlin commented:
Force-pushed because I hit enter too quickly in the GitHub UI when accepting a suggestion

brianhlin committed:
Improve yum retry logic

Co-authored-by: edquist <edquist@users.noreply.github.com>

brianhlin committed:
Remove extraneous -x that doesn't get passed onto the subshell

brianhlin committed:
set -xe explicitly instead of in the shebang in case its sourced

brianhlin committed:
Remove unnecessary subshell

brianhlin committed:
Only create the SOURCES and SPECS dir

rpmbuild creates all of the other dirs for us
